api/src/main/java/org/apache/iceberg/ChangelogScanTask.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+/**
+ * A changelog scan task.
+ */
+public interface ChangelogScanTask extends FileScanTask {
+  /**
+   * Returns the operation type (i.e. insert/delete).
+   */
+  ChangelogOperation operation();
+
+  /**
+   * Returns the relative commit order in which the changes must be applied.
+   */
+  int commitOrder();
+
+  /**
+   * Returns the snapshot ID in which the changes were committed.
+   */
+  long commitSnapshotId();
```

Comment:
Are we restrict this scan task for one snapshot? I feel like itself and its subclass can be used for a scan across multiple snapshots, which is useful, for example, to scan position deletes across snapshots. In that case, commitSnapshotId is going to be optional.

Reply:
When we scan for changes across snapshots, I think we should output all changes per snapshot by default. If I understand you correctly, you want to support use cases when the user is only interested in net changes across snapshots. I think that's possible too but the records produced by a task would still have some snapshot ID when they were committed.

For instance, we add data file A in S1. Then we add a delete file D in S2 that matches data file A. In this case, we would output `AddedDataFileScanTask` with commit snapshot S1 but it will have a delete file D associated with it so that we will only output records that remain there in S2.

Reply:
What will `commitSnapshotId()` return in that case? s1 or s2, or a list of snapshot ids? I'd think the snapshot ids may not be useful if we want to calculate the net changes across snapshots. For example, `IncrementalAppendScan` could be considered as one of cases that outputs scan tasks across snapshots, which doesn't need the snapshot id.

One solution is to add another layer of interface to support the net changes across snapshots. For example, we can use `ChangelogScanTask` as a parent for both cases of one snapshot and multiple snapshots, and then create a new interface `SnapshotChangelogScanTask` to extend it for use case of a single snapshot. 

I think the current solution is good enough for CDC implementation. But, since it is an interface change, we may want to make sure it is extendable for the future use cases. If it is something we will do to support net changes across snapshots, we may choose the right the right hierarchy and names now.

Reply:
After thinking more about this, I agree that commit snapshot ID doesn't make sense for net changes across multiple snapshots in some cases. For instance, when we have 2 snapshots that add delete files for the same data file. In this case, we will have a single task with one data file and two deletes added in multiple snapshots.

Having said that, I am still not sure having separate tasks is a good idea. Apart from more tasks to support in readers, what schema to produce? Our `changelog` metadata table should have the same schema independently whether we include only net changes or all changes per each snapshot.

We can make the commit snapshot ID null when only net changes are produced, skip the commit snapshot ID in both cases or offer some sort of a map of metadata where the snapshot ID may or may not be set.

Let me think more about this.

Reply:
This question is still open. I mainly consider 2 options to handle scans for net changes across snapshots:

- Switch to `Map<String, String> commitMetadata()` that will contain things like snapshot ID and commit timestamp. That would mean having a map instead of separate columns like `_commit_snapshot_id`. It will be then easier to add more info later but it does make it harder to access what's stored in the map.
- Switch to `Long` from `long` and make it null if we are scanning across snapshots.

I'd be curious to hear what other folks think.

Reply:
I may not be completely following, so correct me if I'm not understanding something.

This interface appears to be written to contain changes from a single snapshot. For example, it could be reading deletes that were committed via a delete file that was added in snapshot `S`. In that case, `operation=DELETE`, `changeOrder=ord(S)`, and `commitSnapshotId=S`.

If I understand this thread, the question is how to handle "net changes across snapshots", which I interpret as basically "squashing" snapshots to produce the output. If `S1` adds rows with ids 4, 5, and 6, then `S2` deletes the rows with id 5 and 7, then the result could be the rows with ids 4 and 6, along with a delete for id 7.

We can avoid the problem by always producing one task for `S1` with rows 4, 5, and 6, and then another task (or set of tasks) for `S2` with the delete for ids 5 and 7, which is what the interface appears to be written for. But I think the point is that we can read data files normally and apply deletes to easily produce the squashed version, which could be quite valuable to end users.

I think we can solve this by setting the commit snapshot and change order to the values for `S1`. The difference between the two cases is that the read of the inserted data applies the newer deletes in the window of snapshots. So that task will always produce values that were added in snapshot `S1`, rows 4 and 5. Then a separate task would be created for the deletes from `S2` (with `S2` for its `commitSnapshotId`), which would not emit id 5 because its change order was not strictly less than `ord(S1)`. That task would produce a deleted row for id 7, using the delete task's snapshot ID. So I think in all cases this works. We just still need to separate the deletes in their own task, still.

Reply:
@aokolnychyi, how about creating a new class instead of a `Map<String, String>` for `commitMetadata`? The `commitMetadata` will have `int commitOrder`, `long snapshotId`, `long timestamp`. In case, we have something else needed to pass to reader through the task, we can change the class `commitMetadata` instead of changing this interface.

Reply:
> I think the point is that we can read data files normally and apply deletes to easily produce the squashed version, which could be quite valuable to end users.

Yes. That's one of the use case we think of, which could be resolved by separating the tasks. 

Here is another one, which is a bit different from the above use case, `S1` adds rows with ids 4, 5, and 6, then `S2` deletes the rows with id 5 and 7, `S2` also adds the rows with id 8, 9. We want to have a single `AddedRowsScanTask` to get all added rows with s1 and s2. To give one snapshot id doesn't make sense since added rows 4, 6, 8, 9 come from different snapshots. Of course, we archive the same result by splitting it into two `AddedRowsScanTask`s, and merge them later. But a task across multiple multiple snapshots may have perf advantage, or simpler implementation for that use case. What do you think?

Reply:
Sounds good to me.

Reply:
We won't be able to extend `ChangelogScanTask` in `CombinedDeletedRowsScanTask`, though? Unless we remove the commit snapshot ID from `ChangelogScanTask`?

Reply:
I see your point, but I'm thinking that we would just throw an exception if it were called. I'd also be up for returning null and passing each sub-task's ID and ordinal separately.

Reply:
An exception sounds reasonable. Using a null in a public API is probably not the best idea.